server/server.go

```diff
@@ -0,0 +1,244 @@
+package server
+
+import (
+	"context"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	finderhttpclient "github.com/ipni/storetheindex/api/v0/finder/client/http"
+	"github.com/ischasny/dhfind/metrics"
+)
+
+// preferJSON specifies weather to prefer JSON over NDJSON response when request accepts */*, i.e.
+// any response format, has no `Accept` header at all.
+const preferJSON = true
+
+// simulationWorkerCount is a number of background workers that find tasks are delegated to in simulation mode
+const simulationWorkerCount = 50
+
+var (
+	logger = logging.Logger("server")
+)
+
+type Server struct {
+	s                 *http.Server
+	m                 *metrics.Metrics
+	dhaddr            string
+	simulation        bool
+	simulationJobs    chan *http.Request
+	simulationCancel  context.CancelFunc
+	simulationContext context.Context
+}
+
+// responseWriterWithStatus is required to capture status code from ResponseWriter so that it can be reported
+// to metrics in a unified way
+type responseWriterWithStatus struct {
+	http.ResponseWriter
+	status int
+	// TODO: remove that once the server is run in non-simulaiton mode
+	header http.Header
+}
+
+func newResponseWriterWithStatus(w http.ResponseWriter) *responseWriterWithStatus {
+	ws := &responseWriterWithStatus{
+		ResponseWriter: w,
+		// 200 status should be assumed by default if WriteHeader hasn't been called explicitly https://pkg.go.dev/net/http#ResponseWriter
+		status: 200,
+	}
+	if w == nil {
+		ws.header = make(http.Header)
+	}
+	return ws
+}
+
+func (rec *responseWriterWithStatus) Header() http.Header {
+	if rec.ResponseWriter == nil {
+		return rec.header
+	}
+	return rec.ResponseWriter.Header()
+}
+
+func (rec *responseWriterWithStatus) Write(b []byte) (int, error) {
+	if rec.ResponseWriter == nil {
+		return len(b), nil
+	}
+	return rec.ResponseWriter.Write(b)
+}
+
+func (rec *responseWriterWithStatus) WriteHeader(code int) {
+	rec.status = code
+	if rec.ResponseWriter != nil {
+		rec.ResponseWriter.WriteHeader(code)
+	}
+}
+
+func New(addr, dhaddr string, m *metrics.Metrics, simulation bool) (*Server, error) {
+	var server Server
+
+	server.s = &http.Server{
+		Addr:    addr,
+		Handler: server.serveMux(),
+	}
+	server.m = m
+	server.dhaddr = dhaddr
+	server.simulation = simulation
+	if simulation {
+		server.simulationJobs = make(chan *http.Request, simulationWorkerCount)
```

Comment:
I am curious why the channel buffer is set to worker count. The two seem independent of one another. I recommend refactoring the name or separating channel buffer variable.

Reply:
Good point - it should be unbounded, changing